### PR TITLE
Avoid SITL deadlock when taking semaphores with timeout

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/AP_HAL/Scheduler.h [hunk 1]
```diff
@@ -115,6 +115,12 @@ class AP_HAL::Scheduler {
         return false;
     }
 
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Comment (WickedShell) [inline libraries/AP_HAL/Scheduler.h]: This is messy showing up here for all vehicle. It's a bit less type safe I admit, but I'd rather see this show up as a public member of the HAL_SITL scheduler, and see us cast down where we need it.

**Comment by khancyr** [comment]
Isn't that because SITL clock is updated by the main thread as SITL update is done by the hal ? So if you block the main thread SiTL clock isn't update anymore and everything collapse as you won't get changes in timing anywhere. Did I get it right ? @peterbarker

**Comment by peterbarker** [comment]
@WickedShell Did you see the extended commit comment explaining it?

```
 AP_HAL_SITL: avoid lockstep scheduling issue with semaphores

If there is another thread holding a lock that the main thread wants to
take while pretending to be the IO thread, we will wait forever as we do
not move SITL time forward while pretending to be the IO thread.

This patch simply allows time to move forward if we've failed to take a
semaphore immediately and need to wait.
```

@khancyr has it essentially correct.

**Comment by WickedShell** [comment]
@peterbarker no because that's extended archaeology. Please actually bring these relevant messages into the PR, rather then hoping someone reads every individual commit.

If it's due to the IO stuff being in the main thread this raises the question of why isn't the IO code run in it's own thread?

File: libraries/AP_HAL/Scheduler.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_HAL/Scheduler.h]: It's all in SITL now.

**Pull request merged**